rust_src/src/print.rs

```diff
@@ -0,0 +1,49 @@
+//! Lisp object printing and output
+
+use remacs_macros::lisp_fn;
+
+use crate::{
+    lisp::defsubr,
+    lisp::LispObject,
+    lists::LispCons,
+    remacs_sys::Vprin1_to_string_buffer,
+    remacs_sys::{print_error_message, set_buffer_internal, Fbuffer_string, Ferase_buffer},
+    remacs_sys::{Qerror, Qnil},
+    threads::ThreadState,
+};
+
+/// Convert an error value (ERROR-SYMBOL . DATA) to an error message.
+/// See Info anchor `(elisp)Definition of signal' for some details on how this
+/// error message is constructed.
+#[lisp_fn]
+pub fn error_message_string(obj: LispCons) -> LispObject {
+    let old = ThreadState::current_buffer_unchecked();
+    let value: LispObject;
+
+    /* If OBJ is (error STRING), just return STRING.
+    That is not only faster, it also avoids the need to allocate
+    space here when the error is due to memory full.  */
+    match obj.into() {
+        (error, data) if error.eq(Qerror) => {
+            if let Some((string, nil)) = data.into() {
+                if string.is_string() && nil.is_nil() {
+                    return string;
+                }
+            }
+        }
+        _ => {}
+    }
+
+    unsafe {
+        print_error_message(obj.into(), Vprin1_to_string_buffer, None, Qnil);
```

Comment:
How to give a nullptr for an argument in rust?

Reply:
And I’ｍ wondering about an ownership issue.
at [l.26](https://github.com/Wilfred/remacs/pull/1231/files#diff-78ac0a6803c74006beb627ba7442cdcfR26), 
I use `obj.into()` , which takes the ownership of `obj` and bind it to match arm's expressions.
, then `if let Some(string, nil) = value.into()` also takes that ownership.
Here, if it matches to the `_` arm, or don't get into if-let clause, where `obj`'s ownership go?
and,
Why this reuse obj makes the code complies without error?

Reply:
To pass a null pointer you want to use `std::ptr::null()` or `std::ptr::null_mut()` depending on if you want it to be const or mut.

Reply:
And you should note, as lisp objects are `Copy`, since they're more or less a fancier pointer (that's not the full story but good enough for our purposes), so you generally don't have to worry about ownership issues with them.